hdfs/client.py

```diff
@@ -1049,11 +1049,14 @@ def _walk(dir_path, dir_status, depth):
       if status:
         yield ((dir_path, dir_status), dir_infos, file_infos)
       else:
+        _dir_infos = [name for name, _ in dir_infos]
         yield (
           dir_path,
-          [name for name, _ in dir_infos],
+          _dir_infos,
           [name for name, _ in file_infos],
         )
+        dir_infos_map = dict((name, s) for name, s in dir_infos)
+        dir_infos = [(name, dir_infos_map[name]) for name in _dir_infos]
```

Comment:
It sounds like `os.walk` also supports adding directories. If that's the case, we should probably do the same here (adding logic to fetch any missing status).